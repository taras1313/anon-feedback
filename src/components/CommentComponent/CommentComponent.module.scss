.divider {
  width: 100%;
  background: #dcdcdc;
  height: 1px;
  margin: 20px 0;
}

.flexWrapper {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.comment {
  padding-top: 16px;
}

.text {
  margin: 10px 0;
  line-height: 22px;
  font-size: 15px;
  white-space: pre-line;
}

.leftShift {
  padding-left: 26px;
}

.btnHolder {
  width: 140px;
  display: flex;
  justify-content: space-between;
}