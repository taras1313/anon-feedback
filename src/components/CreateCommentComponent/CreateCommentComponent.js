import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Chip from '@material-ui/core/Chip';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';

import cx from 'classnames';
import { uniqueNamesGenerator, adjectives, animals } from 'unique-names-generator';

const useStyles = makeStyles(theme => ({
  root: {
    width: '80%',
    height: '70vh'
  },
  chip: {
    marginLeft: '6px',
    height: '26px'
  },
  repliedToWrapper: {
    marginTop: '10px'
  },
  appBar: {
    position: 'relative',
    backgroundColor: '#424b5f',
    boxShadow: 'none'
  },
  title: {
    marginLeft: theme.spacing(2),
    textTransform: 'capitalize',
    flex: 1,
  },
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '30px 40px',
    flexDirection: 'column',
    flex: 1
  },
  textField: {
    width: '100%'
  },
  smallTextField: {
    width: 200
  },
  nickNameWrapper: {
    display: 'flex',
    flexDirection: 'column',
    marginTop: '10px'
  },
  formControl: {
    margin: 0
  },
  controlBtns: {
    marginTop: '16px',
    marginLeft: 'auto',
    textTransform: 'capitalize',
    display: 'flex',
    justifyContent: 'flex-end'
  },
  button: {
    fontWeight: 'bold',
    boxShadow: 'none',
    marginLeft: '16px'
  },
  submitBtn: {
    backgroundColor: '#65c178',
    fontWeight: 'bold',
    color: 'white',
    '&:hover': {
      backgroundColor: '#159776'
    },
  }
}));

export const CreateCommentComponent = (
  {
    username = '',
    onCreateComment,
    repliedTo,
    onRepliedToReset
  }
) => {
  const classes = useStyles();
  const [isChecked, setIsChecked] = React.useState(!!username);
  const [nickName, setNickName] = React.useState(username);
  const [commentText, setCommentText] = React.useState('');

  const toggleAutoGenerator = () => {
    const config = {
      dictionaries: [adjectives, animals],
      separator: ' ',
      length: 2,
    };
    
    const value = !isChecked ? uniqueNamesGenerator(config) : '';

    setNickName(value);
    setIsChecked(!isChecked);
  };

  const updateCommentText = ({ target: { value } }) => setCommentText(value);

  const updateUserNickName = ({ target: { value } }) => setNickName(value);

  const resetCommentFields = () => {
    setCommentText('');
    setIsChecked(false);
  };

  const createCommentHandler = () => {
    onRepliedToReset && onRepliedToReset();

    onCreateComment({ nickName, commentText, repliedTo })
      .then(resetCommentFields);
  };

  return (
    <div>
      {!username && (
        <div className={classes.nickNameWrapper}>
          <TextField
            autoComplete='off'
            id="standard-basic"
            className={cx(classes.textField, classes.smallTextField)}
            onChange={updateUserNickName}
            required
            value={nickName}
            name="username"
            label="Anon nick name"
          />
          <FormControlLabel
            className={classes.formControl}
            control={
              <Checkbox
                checked={isChecked}
                onChange={toggleAutoGenerator}
                color="primary"
              />
            }
            label="Generate nick name for you?"
          />
        </div>
      )}

      {repliedTo && (
        <div className={classes.repliedToWrapper}>
          Comment replied to
          <Chip
            label={`@${repliedTo}`}
            onDelete={onRepliedToReset}
            color="primary"
            variant="outlined"
            className={classes.chip}
          />
        </div>
      )}
      <TextField
        required
        autoComplete='off'
        id="standard-multiline-flexible"
        label="Your comment"
        multiline
        rows="6"
        value={commentText}
        onChange={updateCommentText}
        name="text"
        className={classes.textField}
        margin="normal"
        variant="filled" // ?
        // variant="outlined" // ?

      />

      <div className={classes.controlBtns}>
        <Button
          data-testid="here"
          variant="outlined"
          onClick={resetCommentFields}
          className={classes.button}
        >
          Reset
        </Button>

        <Button
          variant="outlined"
          onClick={createCommentHandler}
          color="primary"
          className={cx(classes.button)}
        >
          Create Comment
        </Button>
      </div>
    </div>
  );
};
