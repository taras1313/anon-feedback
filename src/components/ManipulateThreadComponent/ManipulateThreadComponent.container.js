import { connect } from 'react-redux';
import { bindActionCreators, compose } from 'redux';
import { withRouter } from 'react-router-dom';

import { ManipulateThreadComponent as ManipulateThread } from './ManipulateThreadComponent';
import {
  updateThreadField,
  resetThreadFields,
  createThread,
  editThread,
  setSelectedThread
} from '../../actions/threadActions';

const mapStateToProps = ({ threadsReducer: { threadData } }) => ({ threadData });

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators({
    updateThreadField,
    resetThreadFields,
    createThread,
    editThread,
    setSelectedThread
  }, dispatch)
});

export const ManipulateThreadComponent = compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps)
)(ManipulateThread);

