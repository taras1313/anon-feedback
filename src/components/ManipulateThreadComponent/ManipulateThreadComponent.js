import React, { useState, useCallback } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import TextField from '@material-ui/core/TextField';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import CloseIcon from '@material-ui/icons/Close';
import Slide from '@material-ui/core/Slide';
import cx from 'classnames';
import { uniqueNamesGenerator, adjectives, animals } from 'unique-names-generator';

const EDIT = 'edit';
const CREATE = 'create';

const useStyles = makeStyles(theme => ({
  root: {
    width: '80%',
    height: '70vh'
  },
  appBar: {
    position: 'relative',
    backgroundColor: '#424b5f',
    boxShadow: 'none'
  },
  title: {
    marginLeft: theme.spacing(2),
    textTransform: 'capitalize',
    flex: 1,
  },
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '30px 40px',
    flexDirection: 'column',
    flex: 1
  },
  textField: {
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    width: '100%',
    whiteSpace: 'pre-line'
  },
  smallTextField: {
    width: 200
  },
  nickNameWrapper: {
    display: 'flex',
    flexDirection: 'column'
  },
  formControl: {
    margin: 0
  },
  controlBtns: {
    marginTop: '16px',
    marginLeft: 'auto',
    textTransform: 'capitalize'
  },
  button: {
    fontWeight: 'bold',
    boxShadow: 'none',
    marginLeft: '16px',
    fontSize: '0.7rem'
  },
  submitBtn: {
    backgroundColor: '#65c178',
    fontWeight: 'bold',
    color: 'white',
    '&:hover': {
      backgroundColor: '#159776'
    },
  }
}));

const Transition = React.forwardRef(function Transition(props, ref) {
  return <Slide direction="up" ref={ref} {...props} />;
});

export const ManipulateThreadComponent = (
  {
    history,
    action,
    isOpen,
    onClose,
    actions: {
      updateThreadField,
      resetThreadFields,
      createThread,
      editThread,
      setSelectedThread
    },
    threadData: {
      title,
      text,
      username
    }
  }) => {
  const classes = useStyles();
  const [isChecked, setIsChecked] = useState(false);

  const toggleAutoGenerator = () => {
    const config = {
      dictionaries: [adjectives, animals],
      separator: ' ',
      length: 2,
    };

    const value = !isChecked ? uniqueNamesGenerator(config) : '';

    updateThreadField({ field: 'username', value });
    setIsChecked(!isChecked);
  };

  const updateThreadData = ({ target: { value, name } }) => {console.log(value.toString());updateThreadField({ field: name, value })};

  const manipulateThread = () => {
    switch (action) {
      case EDIT: {
        editThread().then(data => {
          setSelectedThread(data);
          onClose();
        });
        break;
      }

      case CREATE: {
        createThread().then(data => {
          const { _id } = data;
          const { push } = history;

          setSelectedThread(data);
          push(`/feed/${_id}`);
          onClose();
        });
        break;
      }

      default:
        return;
    }
  };

  return (
    <div>
      <Dialog
        fullWidth
        maxWidth="md"
        open={isOpen}
        onClose={onClose}
        TransitionComponent={Transition}
      >
        <AppBar className={classes.appBar}>
          <Toolbar>
            <IconButton edge="start" color="inherit" onClick={onClose} aria-label="close">
              <CloseIcon />
            </IconButton>
            <Typography variant="h6" className={classes.title}>
              {action} Thread
            </Typography>
          </Toolbar>
        </AppBar>
        <div className={classes.container}>
          {action === CREATE && (
            <div className={classes.nickNameWrapper}>
              <TextField
                autoComplete='off'
                id="standard-basic"
                className={cx(classes.textField, classes.smallTextField)}
                onChange={updateThreadData}
                required
                value={username}
                name="username"
                label="Anon nick name"
              />
              <FormControlLabel
                className={classes.formControl}
                control={
                  <Checkbox
                    checked={isChecked}
                    onChange={toggleAutoGenerator}
                    color="primary"
                  />
                }
                label="Generate nick name for you?"
              />
            </div>
          )}

          <TextField
            autoComplete='off'
            required
            id="standard-multiline-flexible"
            label="Thread Title"
            rows="2"
            multiline
            value={title}
            onChange={updateThreadData}
            name="title"
            className={classes.textField}
            margin="normal"
            variant="outlined" // ?
            // variant="filled" // ?
          />

          <TextField
            required
            autoComplete='off'
            id="standard-multiline-flexible"
            label="Thread Description"
            multiline
            rows="6"
            value={text}
            onChange={updateThreadData}
            name="text"
            className={classes.textField}
            margin="normal"
            variant="outlined" // ?
            // variant="filled" // ?

          />

          <div className={classes.controlBtns}>
            <Button variant="contained" onClick={resetThreadFields} color="secondary" className={classes.button}>
              Reset
            </Button>

            <Button variant="contained" onClick={manipulateThread} className={cx(classes.button, classes.submitBtn)}>
              {action} Thread
            </Button>
          </div>
        </div>
      </Dialog>
    </div>
  );
};