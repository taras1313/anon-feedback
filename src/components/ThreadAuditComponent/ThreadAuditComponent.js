import React from 'react';
import cx from 'classnames';

import ThumbDownOutlinedIcon from '@material-ui/icons/ThumbDown';
import ThumbUpOutlinedIcon from '@material-ui/icons/ThumbUp';

import styles from './ThreadAuditComponent.module.scss';

export const ThreadAuditComponent = ({
	subscribersCount,
	dislikesCount,
	likesCount,
	isClickable,
	like,
	dislike,
	likeStatus,
	createdDate
}) => {
	return (
		<div className={styles.threadMetaInfo}>
			{subscribersCount} subscribers
			<span className={styles.divider}>-</span>
			{createdDate}
			<span className={styles.divider}>-</span>
			<div onClick={isClickable ? () => like() : null} className={styles.flex}>
				<ThumbUpOutlinedIcon
					className={cx(styles.icon, {
						[styles.like]: likeStatus === 'liked'
					})}
				/>{' '}
				{likesCount}{' '}
			</div>
			<span className={styles.divider}>-</span>
			<div onClick={isClickable ? () => dislike() : null} className={styles.flex}>
				<ThumbDownOutlinedIcon
					className={cx(styles.icon, {
						[styles.dislike]: likeStatus === 'disliked'
					})}
				/>{' '}
				{dislikesCount}{' '}
			</div>
		</div>
	);
};
