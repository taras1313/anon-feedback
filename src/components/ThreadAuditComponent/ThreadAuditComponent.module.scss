@import "../../styles/variables";

.divider {
  margin: 0 8px;
}

.icon {
  max-width: 20px;
  max-height: 20px;
  margin-right: 6px;
}

.threadMetaInfo {
  color: $light-gray;
  display: flex;
  font-size: 13px;
  width: 100%;
  align-items: flex-end;
}