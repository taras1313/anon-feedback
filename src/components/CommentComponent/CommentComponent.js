import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import withStyles from '@material-ui/core/styles/withStyles';
import Chip from '@material-ui/core/Chip';
import cx from 'classnames';

import { ThreadAuthorComponent } from '../ThreadAuthorComponent';
import { ThreadAuditComponent } from '../ThreadAuditComponent';

import styles from './CommentComponent.module.scss';

const classes = () => ({
  chip: {
    marginRight: '6px',
    height: '26px'
  }
});

class Comment extends Component {
  state = {
    isEditing: false,
    text: '',
  };

  toggleIsEditing = () => {
    const { isEditing, text: newCommentText } = this.state;
    const { comment: { text, _id: commentId }, onUpdateComment } = this.props;

    if (!isEditing) { // open editing
      this.setState({ isEditing: !isEditing, text })
    } else { // saving
      onUpdateComment({ text: newCommentText, commentId }).then(() => {
        this.setState({ isEditing: !isEditing, text: '' })
      });
    }
  };

  editComment = ({ target: { value } }) => this.setState({ text: value });

  resetIsEditing = () => this.setState({ isEditing: false });

  renderButtons = () => {
    const { editable } = this.props;
    if (!editable) return null;

    const { isEditing } = this.state;

    return (
      <div className={isEditing ? styles.btnHolder : ''}>
        {isEditing && (
          <Button onClick={this.resetIsEditing} size='small'>
            Cancel
          </Button>
        )}

        <Button onClick={this.toggleIsEditing} color="secondary" size="small">
          {isEditing ? 'Save' : 'Edit'}
        </Button>
      </div>
    );
  };

  constructAdditionalText = () => {
    const { comment: { wasEdited }, isAuthorsComment } = this.props;
    let text = '';
    if (isAuthorsComment) {
      text = '@thread author ';
    }

    if (wasEdited) {
      text += '(edited)'
    }

    return text.trim();
  };

  repliedToHandler = () => {
    const { repliedToHandler, comment: { author: { username } } } = this.props;

    repliedToHandler(username);
  };

  render() {
    const { comment: { text, author: { username }, repliedToUser }, userId, classes } = this.props;
    const { isEditing, text: inputText } = this.state;

    return (
      <div className={styles.comment}>
        <div className={styles.flexWrapper}>
          <ThreadAuthorComponent
            username={username}
            label={this.constructAdditionalText()}
          />
          {this.renderButtons()}
        </div>

        <div className={cx(styles.text, styles.leftShift)}>
          {!!repliedToUser && !isEditing && (
            <Chip
              variant="outlined"
              className={classes.chip}
              label={`@${repliedToUser.username}`}
              color={userId === repliedToUser.userId ? 'secondary' : 'primary'}
            />
          )}
          {!isEditing && text}

          {isEditing && (
            <TextField
              autoComplete='off'
              id="standard-multiline-flexible"
              label="Edit your comment"
              multiline
              value={inputText}
              fullWidth
              onChange={this.editComment}
              // margin="normal"
              variant="outlined"
              // variant="outlined" // ?
            />
          )}
        </div>

        <div className={cx(styles.flexWrapper, styles.leftShift)}>
          <ThreadAuditComponent />
          <Button color='primary' onClick={this.repliedToHandler} variant='outlined' size='small'>
            Reply
          </Button>
        </div>
        <div className={styles.divider} />
      </div>
    );
  }
}

export const CommentComponent = withStyles(classes)(Comment);