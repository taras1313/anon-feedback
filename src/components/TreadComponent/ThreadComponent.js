import React, { Component, Fragment, useState } from 'react';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import QuestionAnswerOutlinedIcon from '@material-ui/icons/QuestionAnswerOutlined';
import cx from 'classnames';

import styles from './ThreadComponent.module.scss';
import { ThreadAuthorComponent } from '../ThreadAuthorComponent';
import { ThreadAuditComponent } from '../ThreadAuditComponent';
import { ManipulateThreadComponent } from '../ManipulateThreadComponent';
import { CommentComponent } from '../CommentComponent';
import { CreateCommentComponent } from '../CreateCommentComponent';
import Loader from '../Loader/Loader';

const randomize = () => Math.floor(Math.random() * 10);
const PREVIEW = 'preview';
const FULL = 'full';
const viewTypes = [PREVIEW, FULL];

//todo like logic- when to disable, already liked functionality, dislike cancels like etc.

export class ThreadComponent extends Component {
  constructor(props) {
    super(props);
    const { threadView } = props;

    this.renderContent = threadView === FULL ? this.renderThread : this.renderPreviewThread;

    this.state = {
      editThreadOpen: false
    };
  }

  closeEditThread = () => this.setState({ editThreadOpen: false });

  openEditThread = () => {
    const {
      actions: { setThreadData },
      thread: { text, _id, title }
    } = this.props;
    setThreadData({ text, _id, title });

    this.setState({ editThreadOpen: true });
  };

  renderPreviewThread = () => {
    const {
      thread: {
        author: { username } = {},
        answersCount,
        subscribersCount,
        dislikesCount,
        likesCount
      }
    } = this.props;

    return (
      <Fragment>
        <div>
          <ThreadAuthorComponent username={username} />
          {this.renderThreadHeader()}
          <ThreadAuditComponent
            subscribersCount={subscribersCount}
            dislikesCount={dislikesCount}
            likesCount={likesCount}
						likeStatus={this.likeStatus()}
					/>
        </div>
        <div className={styles.answersCount}>
          <div className={styles.counterWrapper}>
            <QuestionAnswerOutlinedIcon />
            {answersCount}
          </div>
          <span>answer(s)</span>
        </div>
      </Fragment>
    );
  };

  subscriptionHandler = () => {
    const {
      actions: {
        subscribeToThread,
        unsubscribeFromThread
      },
      thread: { _id: threadId },
      userId
    } = this.props;

    const isSubscribed = this.isSubscribed();
    const params = { threadId, userId };

    if (isSubscribed) {
      unsubscribeFromThread(params)
    } else {
      subscribeToThread(params);
    }
  };

  isSubscribed = () => {
    const { thread: { subscribers }, userId } = this.props;
    return !!subscribers.find(user => user.userId === userId);
  };

  isAuthor = () => {
    const { thread: { author: { userId: authorId } }, userId } = this.props;

    return authorId === userId;
  };

  findUserInThread = () => {
    const { userId, thread: { users } } = this.props;

    return users.find(user => user.userId === userId);
  };

	likeStatus = () => {
		const {
			userId,
			thread: {
				likesList,
				dislikesList
			}
		} = this.props;

		if (likesList.find(el => el.userId === userId)) return 'liked';
		if (dislikesList.find(el => el.userId === userId)) return 'disliked';
		return null
	};

  renderThread = () => {
    const { editThreadOpen } = this.state;

    const {
      userId,
      thread: {
        text,
        subscribersCount,
        dislikesCount,
        likesCount,
        commentsList,
        author: { userId: threadAuthorId, username: authorName } = {}
      },
			likeThread,
			dislikeThread,
      actions: { onCreateComment, onUpdateComment }
    } = this.props;
    const subscribed = this.isSubscribed();
    const { username } = this.findUserInThread() || {};

    return (
      <div className={styles.threadFullWrapper}>
        <div className={styles.headingWrapper}>
          <ThreadAuthorComponent username={authorName} />
          {
            this.isAuthor() && <Button onClick={this.openEditThread} color="secondary">
              Edit
            </Button>
          }
        </div>
        <div className={styles.divider} />
        {this.renderThreadHeader()}
        <div className={styles.description}>
          {text}
        </div>
        <div className={styles.subsWrapper}>

          <ThreadAuditComponent
            subscribersCount={subscribersCount}
            dislikesCount={dislikesCount}
            likesCount={likesCount}
						likeStatus={this.likeStatus()}
						dislike={dislikeThread.bind(null, {id, userId})}
						like={likeThread.bind(null, {id, userId})}
						isClickable
          />

          <Button
            className={styles.btn}
            variant="outlined"
            onClick={this.subscriptionHandler}
            color={subscribed ? 'secondary' : 'primary'}
          >
            {subscribed ? 'Unsubscribe' : 'Subscribe'}
          </Button>
        </div>
        <div className={styles.dividerWrapper}>
          Total comments ({8})
        </div>

        <div className={styles.comments}>
          {commentsList.map(el => {

            return (
              <CommentComponent
                onUpdateComment={onUpdateComment}
                editable={el.author.userId === userId}
                comment={el}
                isAuthorsComment={el.author.userId === threadAuthorId}
              />
            )
          })}
        </div>

        <div className={styles.dividerWrapper}>
          {!!username ?
            <span className={styles.userHighlight}>{username},</span> : ''} Write your comment here
        </div>
        <CreateCommentComponent onCreateComment={onCreateComment} username={username} />

        <ManipulateThreadComponent
          onClose={this.closeEditThread}
          isOpen={editThreadOpen}
          action="edit"
        />
      </div>
    );
  };

  renderThreadHeader = () => {
    const { threadView } = this.props;
    const { title = 'What to learn SWIFT or GO?', _id: id } = this.props.thread;

    return (
      <NavLink to={`/feed/${id}`}>
        <div
          className={cx(styles.heading, {
            [styles.hovered]: threadView === PREVIEW
          })}
        >
          {title}
        </div>
      </NavLink>
    );
  };


  render() {
    return <div className={styles.wrapper}>{this.renderContent()}</div>;
  }
}

ThreadComponent.propTypes = {
  heading: PropTypes.string,
  answersCount: PropTypes.number,
  subscribersCount: PropTypes.number,
  creationDate: PropTypes.object,
  content: PropTypes.string,
  likesCount: PropTypes.number,
  dislikesCount: PropTypes.number,
  threadView: PropTypes.oneOf(viewTypes),
  author: PropTypes.object,
  onLikeClick: PropTypes.func,
  onDislikeClick: PropTypes.func,
  onSubscribeClick: PropTypes.func,
  onEditClick: PropTypes.func
};

ThreadComponent.defaultProps = {
  threadView: FULL, // edit
  subscribersCount: randomize(),
  answersCount: randomize(),
  dislikesCount: randomize(),
  likesCount: randomize(),
  author: {}
};
