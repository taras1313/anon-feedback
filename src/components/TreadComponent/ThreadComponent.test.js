import React from 'react';
import { ThreadComponent } from './ThreadComponent';
import { CreateCommentComponent } from '../CreateCommentComponent';
import { CommentComponent } from '../CommentComponent';
import { ManipulateThreadComponent } from '../ManipulateThreadComponent';
import { ThreadAuditComponent } from '../ThreadAuditComponent';
import { shallow } from 'enzyme';

window.scrollTo = jest.fn();

describe('Thread Component', () => {
	const testAuthor = { userId: 1 };
	const props = {
		userId: 1,
		thread: {
			users: [
				{
					userId: 1
				}
			],
			likesList: [{ userId: 1 }],
			commentsList: [{ text: 'test', author: testAuthor }],
			subscribers: [{ userId: 1 }],
			author: testAuthor
		}
	};

	afterEach(() => {
		jest.clearAllMocks();
	});

	test('renders properly with no props', () => {
		const wrapper = shallow(<ThreadComponent threadView="preview" />);

		expect(wrapper).toMatchSnapshot();
	});

	test('renders properly with no props', () => {
		const wrapper = shallow(<ThreadComponent />);
		
		expect(wrapper).toMatchSnapshot();
	});
	
	test('default thread likes', () => {
		const wrapper = shallow(<ThreadComponent />);

		 wrapper.find(ThreadAuditComponent).props().like();
		 wrapper.find(ThreadAuditComponent).props().dislike();
	});

	test('subscribe to thread works', () => {
		const subscribeToThread = jest.fn();

		const wrapper = shallow(
			<ThreadComponent
				actions={{ subscribeToThread: subscribeToThread }}
				thread={{ ...props.thread, dislikesList: [{ userId: 1 }] }}
			/>
		);

		wrapper.find("[data-testid='subscribeButton']").simulate('click');
		expect(subscribeToThread).toHaveBeenCalledTimes(1);
	});

	test('unsubscribe from thread works', () => {
		const unsubscribeFromThread = jest.fn();
		const wrapper = shallow(
			<ThreadComponent {...props} actions={{ unsubscribeFromThread: unsubscribeFromThread }} />
		);

		wrapper.find("[data-testid='subscribeButton']").simulate('click');
		expect(unsubscribeFromThread).toHaveBeenCalledTimes(1);
	});

	test('thread edit', () => {
		const unsubscribeFromThread = jest.fn();
		const wrapper = shallow(
			<ThreadComponent
				{...props}
				actions={{ unsubscribeFromThread: unsubscribeFromThread, setThreadData: jest.fn() }}
			/>
		);

		wrapper.find("[data-testid='editThread']").simulate('click');
		expect(wrapper.state().editThreadOpen).toBe(true);
		wrapper.find(ManipulateThreadComponent).props().onClose();
		expect(wrapper.state().editThreadOpen).toBe(false);
	});

	test('replied to', () => {
		const wrapper = shallow(<ThreadComponent {...props} />);

		wrapper.find("[data-testid='comment']").props().repliedToHandler('test');
		expect(wrapper.state().repliedTo).toBe('test');
	});

	test('comment', () => {
		const wrapper = shallow(<ThreadComponent {...props} />);

		wrapper.find(CreateCommentComponent).props().onRepliedToReset();
		wrapper.find(CommentComponent).props().likeComment();
		wrapper.find(CommentComponent).props().dislikeComment();

		expect(wrapper.state().repliedTo).toBe('');
	});
});
