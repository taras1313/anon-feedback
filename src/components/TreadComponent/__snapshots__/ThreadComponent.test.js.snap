// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thread Component renders properly with no props 1`] = `
<div
  className="wrapper"
>
  <div>
    <ThreadAuthorComponent />
    <NavLink
      to="/feed/undefined"
    >
      <div
        className="heading hovered"
      >
        What to learn SWIFT or GO?
      </div>
    </NavLink>
    <ThreadAuditComponent
      likeStatus={null}
    />
  </div>
  <div
    className="answersCount"
  >
    <div
      className="counterWrapper"
    >
      <QuestionAnswerOutlinedIcon />
    </div>
    <span>
      answer(s)
    </span>
  </div>
</div>
`;

exports[`Thread Component renders properly with no props 2`] = `
<div
  className="wrapper"
>
  <div
    className="threadFullWrapper"
  >
    <div
      className="headingWrapper"
    >
      <ThreadAuthorComponent />
      <WithStyles(ForwardRef(Button))
        color="secondary"
        data-testid="editThread"
        onClick={[Function]}
      >
        Edit
      </WithStyles(ForwardRef(Button))>
    </div>
    <div
      className="divider"
    />
    <NavLink
      to="/feed/undefined"
    >
      <div
        className="heading"
      >
        What to learn SWIFT or GO?
      </div>
    </NavLink>
    <div
      className="description"
    />
    <div
      className="subsWrapper"
    >
      <ThreadAuditComponent
        dislike={[Function]}
        isClickable={true}
        like={[Function]}
        likeStatus={null}
      />
      <WithStyles(ForwardRef(Button))
        className="btn"
        color="primary"
        data-testid="subscribeButton"
        onClick={[Function]}
        variant="outlined"
      >
        Subscribe
      </WithStyles(ForwardRef(Button))>
    </div>
    <div
      className="dividerWrapper"
    >
      Total comments (
      )
    </div>
    <div
      className="comments"
    />
    <div
      className="dividerWrapper"
    >
       Write your comment here
    </div>
    <CreateCommentComponent
      onRepliedToReset={[Function]}
      repliedTo=""
    />
    <withRouter(Connect(ManipulateThreadComponent))
      action="edit"
      isOpen={false}
      onClose={[Function]}
    />
  </div>
</div>
`;
